src/mlpack/methods/cf/neighbor_search_policies/pearson_search.hpp

```diff
@@ -0,0 +1,105 @@
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_CF_PEARSON_SEARCH_HPP
+#define MLPACK_METHODS_CF_PEARSON_SEARCH_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/methods/neighbor_search/neighbor_search.hpp>
+
+namespace mlpack {
+namespace cf {
+
+/**
+ * Nearest neighbor search with pearson distance (or furthest neighbor search
+ * with pearson correlation).
+ * Note that, with normalized vectors, neighbor search with pearson distance is
+ * equivalent to neighbor search with Euclidean distance. Therefore, instead
+ * of performing neighbor search directly with pearson distance, we first
+ * normalize all vectors, and then use neighbor::KNN (i.e. NeighborSearch with
+ * Euclidean distance, KDTree). Pearson correlation are calculated from
+ * Euclidean distance.
+ */
+class PearsonSearch
+{
+ public:
+  /**
+   * Constructor with reference set.
+   * In order to use neighbor::KNN(i.e. NeighborSearch with Euclidean distance,
+   * KDTree), we need to normalize all vectors in referenceSet. For each vector
+   * x, we first subtract mean(x) from each element in x. Then, we normalize
+   * the vector to unit length.
+   *
+   * @param Set of reference points.
+   */
+  PearsonSearch(const arma::mat& referenceSet)
+  {
+    // Normalize all vectors in referenceSet.
+    // For each vector x, first subtract mean(x) from each element in x.
+    // Then normalize the vector to unit length.
+    arma::mat normalizedSet(arma::size(referenceSet));
+    for (size_t i = 0; i < referenceSet.n_cols; i++)
+    {
+      // Subtract mean(x) from each element in x.
+      normalizedSet.col(i) =
+          referenceSet.col(i) - arma::mean(referenceSet.col(i));
+      // Normalize the vector to unit length.
+      normalizedSet.col(i) = arma::normalise(normalizedSet.col(i), 2);
+    }
+
+    neighborSearch.Train(std::move(normalizedSet));
+  }
+
+  /**
+   * Given a set of query points, find the nearest k neighbors, and return
+   * similarities. Similarities are non-negative and no larger than one.
+   *
+   * @param query A set of query points.
+   * @param k Number of neighbors to search.
+   * @param neighbors Nearest neighbors.
+   * @param similarites Similarities between query point and its neighbors.
+   */
+  void Search(const arma::mat& query, const size_t k,
+      arma::Mat<size_t>& neighbors, arma::mat& similarities)
+  {
+    // Normalize all vectors in query.
+    // For each vector x, first subtract mean(x) from each element in x.
+    // Then normalize the vector to unit length.
+    arma::mat normalizedQuery(arma::size(query));
+    for (size_t i = 0; i < query.n_cols; i++)
+    {
+      // Subtract mean(x) from each element in x.
+      normalizedQuery.col(i) = query.col(i) - arma::mean(query.col(i));
```

Comment:
It should be possible move this one outside the loop and write:

```c++
normalizedQuery = query - arma::mean(query);
```

I think the same is true for the normalize step.